src/clp_ffi_py/ir/native/PyDeserializer.cpp

```diff
@@ -0,0 +1,297 @@
+    if (nullptr == m_deserializer_buffer_reader) {
+        return false;
+    }
+
+    try {
+        PyDeserializer::IrUnitHandler::LogEventHandle log_event_handle
+                = [this](clp::ffi::KeyValuePairLogEvent&& kv_log_event) -> IRErrorCode {
+            return this->handle_log_event(std::move(kv_log_event));
+        };
+
+        PyDeserializer::IrUnitHandler::UtcOffsetChangeHandle trivial_utc_offset_handle
+                = []([[maybe_unused]] clp::UtcOffset, [[maybe_unused]] clp::UtcOffset
+                  ) -> IRErrorCode { return IRErrorCode::IRErrorCode_Success; };
+
+        PyDeserializer::IrUnitHandler::SchemaTreeNodeInsertionHandle
+                trivial_schema_tree_node_insertion_handle
+                = []([[maybe_unused]] clp::ffi::SchemaTree::NodeLocator) -> IRErrorCode {
+            return IRErrorCode::IRErrorCode_Success;
+        };
+
+        PyDeserializer::IrUnitHandler::EndOfStreamHandle end_of_stream_handle
+                = [this]() -> IRErrorCode { return this->handle_end_of_stream(); };
+
+        auto deserializer_result{Deserializer::create(
+                *m_deserializer_buffer_reader,
+                {std::move(log_event_handle),
+                 std::move(trivial_utc_offset_handle),
+                 std::move(trivial_schema_tree_node_insertion_handle),
+                 std::move(end_of_stream_handle)}
+        )};
+        if (deserializer_result.has_error()) {
+            PyErr_Format(
+                    PyExc_RuntimeError,
+                    cDeserializerCreateErrorFormatStr.data(),
+                    deserializer_result.error().message().c_str()
+            );
+            return false;
+        }
+        m_deserializer = new clp::ffi::ir_stream::Deserializer<PyDeserializer::IrUnitHandler>{
+                std::move(deserializer_result.value())
+        };
+    } catch (clp::TraceableException& exception) {
+        handle_traceable_exception(exception);
+        return false;
+    }
+
+    return true;
+}
+
+auto PyDeserializer::deserialize_log_event() -> PyObject* {
+    try {
+        while (false == is_stream_completed()) {
+            auto const ir_unit_type_result{
+                    m_deserializer->deserialize_next_ir_unit(*m_deserializer_buffer_reader)
+            };
+            if (ir_unit_type_result.has_error()) {
+                auto const err{ir_unit_type_result.error()};
+                if (std::errc::result_out_of_range != err) {
+                    PyErr_Format(
+                            PyExc_RuntimeError,
+                            cDeserializerDeserializeNextIrUnitErrorFormatStr.data(),
+                            err.message().c_str()
+                    );
+                    return nullptr;
+                }
+                if (false == handle_incomplete_stream_error()) {
+                    return nullptr;
+                }
+                break;
```

Comment:
what if we simply do this and remove the handle_incomplete_stream_error() method? The error handling should be the same but there will be a difference on how we report the error, not sure if it is an issue.
```suggestion
                if (std::errc::result_out_of_range == err && m_allow_incomplete_stream) {
                    handle_end_of_stream();
                    break;
                }
                PyErr_Format(
                    PyExc_RuntimeError,
                    cDeserializerDeserializeNextIrUnitErrorFormatStr.data(),
                    err.message().c_str()
                );
                return nullptr;
```
```

Reply:
I'm afraid the error reporting will be an issue:
- Currently we're using some std::error_code to indicate IR errors (no our own error code enums until we get this PR merged: https://github.com/y-scope/clp/pull/486). Using these error codes won't give users a clear error message on incomplete IR stream
- For incomplete IR stream, we should set the exception to `PyDeserializerBuffer::get_py_incomplete_stream_error()` which is a customized CPython exception specific to incomplete stream error.

Reply:
Fair enough, let's keep ur current version